helix-core/src/main/java/org/apache/helix/controller/VirtualTopologyGroupStrategy.java

```diff
@@ -0,0 +1,80 @@
+package org.apache.helix.controller;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.util.HelixUtil;
+
+public enum VirtualTopologyGroupStrategy {
```

Comment:
Usually Java Enums are used to define values where Java Interfaces are used to define behavior (or the default behavior and make it extendible). Is there a reason we're making this an Enum here?

What I would suggest is to create a separate package (I'm not sure this belongs to `org.apache.helix.controller` - maybe something like `*.cloud.groupingStrategy` or `*.topology.groupingStrategy` might be more appropriate?) and define an interface with a default implementation?

Reply:
Thanks for the comment, it's really good question! I hesitated a lot in choosing enum over interface, here are my considerations:
1. Enum defines constants that could go with predefined behavior (enum methods), and strategy falls in the category. Plus enum is naturally static final singleton, and it's compile-time constant, method is also resolved at compile time, while interface is not.
2. Afaik, we don't have dependency injection to decouple interface and impl, instead we directly construct impl class. I would try to avoid creating interface if there is only one known impl.
3. The key method is static. With interface we could use a static or default method, but it's weird to instantiate a class for static method. 
4. Writing logic in a util class also works, while enum provides a layer of encapsulation and is type safe.

This is open discussion, using interface here won't be too bad after all. Let me know how you feel about it. Thanks again.

Reply:
Could `VirtualTopologyGroupStrategy` have multiple implementations and do we expect to add different implementations and make this configurable? If so, I think using an abstract base class could also work.

Or is the goal to provide a static method? if so i think a util class would fit better here. Since this is only used in `VirtualTopologyGroupService` and only fixed to a `DEFAULT` implementation. Is this class only going to be used in Helix REST? If that's the case, I'd place this in that module.

I'm confused because I'm not sure what the intent of this class is - it also has an abstract method and the default enum is a behavior (verb) than named variables (nouns) that enums generally are.

Reply:
The Enum `VirtualTopologyGroupStrategy` may have other enum values, e.g. `STRATEGY1`, `STRATEGY2` (in parallel to `DEFAULT` , they can have their own method implementation. And it can be configurable which one to use by parsing string.
To clarify, the reason I created this class is to have a way to "implement" static method as in interface. So each enu value defines a strategy, while each strategy comes with an assignment algorithm (abstract method). Indeed, a util class fits here, but I chose to use Enum for better encapsulation. FYI, https://blogs.oracle.com/javamagazine/post/how-to-make-the-most-of-java-enums. 
That said, I'm afraid this is creating more confusion than I expected, I'll change it to a util class if you don't have objection. We can revisit this when we want to add other implementations. Thanks.